### PR TITLE
T/121: Extract the `transformCommit` function from writer-options.js

### LINKED ISSUE #121: [ENV] Extract the `transformCommit` function from writer-options.js
In order to do #110, we need to extract the `transformCommit()` function from [`writer-options.js`](https://github.com/ckeditor/ckeditor5-dev/blob/master/packages/ckeditor5-dev-env/lib/release-tools/changelog/writer-options.js) file.

It allows replacing rules which are used to generate the changelog using parameters.

At this moment, the function is hardcoded and we aren't able to introduce other rules, e.g. dev packages (`ckeditor5-dev-*`) needs other rules than editor's packages (`ckeditor5-*`).

### PR DESCRIPTION
**Pull request by pomek**
Internal: Moved `transformCommit` function from `writer-options` to new file. Closes #121.

It allows replacing rules which are used to generate the changelog using parameters.

### CODE REVIEW

**Comment by pomek** [comment]
In this PR I corrected a little the files structure.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommit.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
+/**
+ * @license Copyright (c) 2003-2017, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+const getPackageJson = require( './getpackagejson' );
+
+const transformCommit = {
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommit.js]: 1. The file name `transformCommit` is suggesting, that the default export should be a function, which is not. 
2. This file name should contains `-` between words according to our style guide :D

Maybe `transform-commit-utils`/ `transform-commit-helpers`?

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommit.js]: :+1: for `transform-commit-utils`

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js [hunk 2]
```diff
@@ -14,7 +15,7 @@ const { commitTypes, transform } = require( '../changelog/writer-options' );
  */
 module.exports = function getNewReleaseType() {
 	const conventionalRecommendedBump = require( 'conventional-recommended-bump' );
-	const parserOpts = require( '../changelog/parser-options' );
+	const parserOpts = require( './parseroptions' );
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: Maybe `parserOptions`? You're saving 4 chars ;)
And now I don't know whether the file name is correct :D It's a plain object, so probably it shouldn't contain '-'.

Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: Ow. Now I see, that there is such a convention with `opts` across the committed files.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js [hunk 3]
```diff
@@ -5,7 +5,8 @@
 
 'use strict';
 
-const { commitTypes, transform } = require( '../changelog/writer-options' );
+const { availableTypes } = require( './transformcommit' );
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: The name `availableTypes` becomes too generic when it's used without the `transformCommit` context.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	for ( const note of commit.notes ) {
+		if ( note.title === 'BREAKING CHANGE' ) {
+			note.title = 'BREAKING CHANGES';
+		}
+		note.text = transformCommit.linkGithubIssues( transformCommit.linkGithubUsers( note.text ) );
+	}
+
+	// Removes references that already appear in the subject.
+	commit.references = commit.references.filter( ( reference ) => {
+		return issues.indexOf( reference.issue ) === -1;
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: It can be replaced with `includes()`

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+	} );
+
+	return commit;
+};
+
+/**
+ * @typedef {Object} Commit
+ *
+ * @property {String} [type] Type of the commit.
```
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: Should every property in type Commit be optional?

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: Yes, if commit does not match to our convention, these fields won't have a value.

**Comment by pomek** [comment]
Fixed. Review requested.

**Comment by pomek** [comment]
I found one issue with the tool, so please do not merge this PR.

**Comment by pomek** [comment]
Eh. I cannot reproduce it. Everything works fine so it could be something on my side.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js [hunk 6]
```diff
@@ -5,7 +5,8 @@
 
 'use strict';
 
-const { commitTypes, transform } = require( '../changelog/writer-options' );
+const { availableCommitTypes } = require( './transform-commit-utils' );
+const transformCommitForCkeditor5Package = require( './transformcommitforckeditor5package' );
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: Another rule – it's `...ForCKEditor5Package`

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: Fixed in .5326cae

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+ * Parses a single commit:
+ *   - displays a log when the commit has invalid format of the message,
+ *   - filters out the commit if it should not be visible in the changelog,
+ *   - makes links to issues and user's profiles on GitHub.
+ *
+ * @param {Commit} commit
+ * @param {Boolean} [displayLog=true] Whether to display the logs.
+ * @returns {Commit}
+ */
+module.exports = function transformCommitForCkeditor5Package( commit, displayLog = true ) {
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: Couldn't this function be called simpler? More general? Like `transformCommit`? ;) Do you expect to have different implementations in the future? With different rules?

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: Yes, for `ckeditor5-dev` we will have another `transformCommit` function.

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: To be honest, the whole refactoring is exactly for it.

File: packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js [hunk 8]
```diff
@@ -66,10 +66,11 @@ module.exports = function generateChangelog( newVersion = null ) {
 					merges: undefined,
 					firstParent: true
 				};
-				const parserOpts = require( '../changelog/parser-options' );
-				const writerOpts = require( '../changelog/writer-options' );
+				const parserOptions = require( '../utils/parser-options' );
+				const writerOptions = require( '../utils/writer-options' );
+				writerOptions.transform = require( '../utils/transformcommitforckeditor5package' );
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js]: You must not touch an object which was returned to you by reference. This way you're affecting someone else's `writerOptions` too. So either make `parser-options` and `writer-options` functions which return a new instance every time or clone the object here.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	if ( hasCorrectType && isCommitIncluded ) {
+		logMessage += chalk.green( 'INCLUDED' );
+	} else if ( hasCorrectType && !isCommitIncluded ) {
+		logMessage += chalk.grey( 'SKIPPED' );
+	} else {
+		logMessage += chalk.red( 'INVALID' );
+	}
+
+	if ( commit.header.startsWith( 'Merge' ) && hasCorrectType ) {
+		const indentSize = '[XX:YY:ZZ] * 1234567 '.length;
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: What's this?

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js [hunk 10]
```diff
@@ -5,7 +5,8 @@
 
 'use strict';
 
-const { commitTypes, transform } = require( '../changelog/writer-options' );
+const { availableCommitTypes } = require( './transform-commit-utils' );
+const transformCommitForCKEditor5Package = require( './transformcommitforckeditor5package' );
```
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: Since a specific "CKEditor 5 package" related "transform commit" function is used here... isn't this function related to "CKEditor 5 package" too?

Wouldn't it make sense if `transformCommit` was this function's param?

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 7 above)
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: OK, I wanted to be sure.

Then, the second question is whether we expect to have more functions related to "CKE 5 packages" and "CKE 5 dev packages"? If so, then perhaps we should move this code somewhere else. Keep it closer to each other. E.g. the tasks available in `index.js` are also CKE5 related only so soon we'll have "ForCKEditor5Package" everywhere. Hence, I wanted to propose to try to better organise this code. But I don't know it too well, so nothing concrete.

File: packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js (hunk 8 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/tasks/generatechangelog.js]: Fixed in d9deb9d.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 9 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: This is a size of indent used in changelog tools. 

Changelog prints out the commits which will be used to generate changelog. For merge commit, we displayed "Merge ..." what was unreadable.

I added an additional note to the second line (below the "Merge ..." message) but it does not look good, so I needed to add the indent. 

I just could put the 21 spaces but nobody understands what I did it.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js (hunk 10 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: In the future, this function will receive a parameter which allows choosing which "transform" function will be used. 

At this moment we have only one transform, so the parameter is not required.

OTOH, I don't know whether the parameter will have a sense. Maybe you are right now.

ATM I have a "YAGNI" feeling.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 9 above)
Comment (ma2ciek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: maybe it could be `INDENT_SIZE = '[XX:YY:ZZ] * 1234567 '.length // 21` at the head of this file (?)

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 7 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: > Then, the second question is whether we expect to have more functions related to "CKE 5 packages" and "CKE 5 dev packages"?

Is the question - do we want to use this tool for generating the changelog for the main (`ckeditor5`) package?

I would like to wait for the second "transformCommit" function. In the next PR (which will introduce release tool for this package) I will propose even better the files structure.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 9 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: I moved the `INDENT_SIZE` at the beginning of the file. Thanks, @ma2ciek.

I added also a short comment which explains what it this.

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: Fixed in 2bd2a63.

Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: The value of this string was dragging my entire attention so I didn't even check what's after it. Code should be minimal also in this sense – that it doesn't contain distractions like such totally unnecessary values.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js (hunk 10 above)
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: So, if you have YAGNI now, then this entire refactoring may be YAGNI ;). It's the same category of problem - do we need to parametrize the `transformCommit`?

But, it'd also be more clear whether that specific function can be used directly here if the code was structured more clearly. As I mentioned before, you should physically separate it into general release helpers, tasks and helpers for CKEditor 5 packages and tasks and helpers for CKEditor 5 dev stuff. You'd then better see where this function landed. If in the first category, it should not know about `transformCommitForCKEditor5Package`.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js (hunk 7 above)
Comment (Reinmar) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/transformcommitforckeditor5package.js]: In such case, this refactoring shouldn't be a separate PR because we're not able to review it. It's a WIP – a state of the code in the middle of a refactoring process.

We can merge it like it looks now but it'd be much easier to make decisions if we've seen the entire picture.

File: packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js (hunk 10 above)
Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: >  do we need to parametrize the transformCommit?

How do you want to generate changelog for the `ckeditor5-dev` package?

Because we talked about it via Slack, I know what we need to do. If you have a batter idea, let me know.

Comment (pomek) [inline packages/ckeditor5-dev-env/lib/release-tools/utils/getnewreleasetype.js]: As you wanted, I've added a parameter to this function.

**Comment by pomek** [comment]
I'm closing this PR. As you mentioned, this is some part of the whole mechanism.

I will create another PR when the whole mechanism will be ready.

**Comment by ma2ciek** [comment]
Sorry @pomek for reviewing the closed stuff ;)
